flesh out SparkBackend and add some simple tests

catoverdrive opened the pull request:
In `SparkBackend`, TableIRs get lowered into SparkStages and value IRs get lowered into SparkPipelines.

`SparkStage` represents the necessary computation on a partition of a table, as well as the partitioning information. This can either directly represent a TableIR, in which case the partition IR (`body`) is an array of all the rows of that given partition, or whatever the downstream ValueIR needs---e.g. for `TableCount`, the length of that array; for `TableWrite`, the filename that the partition was written out to, etc.

`SparkPipeline` represents a local value that can use the results from the referenced stages.

One assumption that I've made in this PR is that none of the bindings across all `SparkStage.globals` will have the same name, and none of them will be named "context". (I think this is a fairly reasonably assumption, since we'll just use genUID() to generate unique IDs for all of them and then use unique symbols once #5080 goes in.)

In this PR, I've lowered:
- TableCount
- TableCollect
- TableGetGlobals
- TableRange
- TableMapGlobals
- TableMapRows

in order to write some tests.

catoverdrive commented:
cc @cseed

cseed commented on hail/src/test/scala/is/hail/backend/spark/SparkBackendSuite.scala:
```diff
@@ -0,0 +1,43 @@
+
+  @Test def testRangeCount() {
+    val node = ir.ApplyBinaryPrimOp(ir.Add(), ir.TableCount(ir.TableRange(10, 2)), ir.TableCount(ir.TableRange(15, 5)))
+    assert(SparkBackend.execute(hc.sc, node, optimize = false) == 25)
```
So these should be just be TableIR tests in TableIRSuite.  You should call assertEvalsTo, etc., but those functions should run the SparkBackend (when it implements the pipeline) and compare results.  We don't want to duplicate the set of IR functionality for each backend, but have a broad set of tests that verify standard behavior and all backends should be tested by those tests.

You could also have testing functions for TableIR that collects the results and compares them.

cseed commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -1,29 +1,209 @@
+    val gEncoder = codecSpec.buildEncoder(gType.physicalType)(baos)
+    gEncoder.writeRegionValue(region, rvb2.end())
+    gEncoder.flush()
+    val globsBC = baos.toByteArray
```
I think this can be replaced by `RegionValue.toBytes`, and analogously `fromBytes` for the decoder.

catoverdrive commented:
Ok, I've moved the tests from their own test suite into TableIR, and adjusted the `assertEvalsTo` stuff to go through `SparkBackend`. 

@danking I had to add `ArrayFlatMap` to cxx.Emit in order to get the relevant tests to start actually getting compiled through the SparkBackend. It's a pretty straightforward modification of the ArrayMap logic above it, and it's being tested through all the IRSuite tests, but if you'd like me to break it out, please let me know.

catoverdrive commented on hail/src/main/scala/is/hail/cxx/Compile.scala:
```diff
@@ -31,7 +31,7 @@ object Compile {
       "char *")
 
-    val v = Emit(fb, 1, body)
+    val v = Emit(fb, 1, ir.Subst(body, ir.Env(arg0 -> ir.In(0, arg0Type.virtualType))))
```
I think this was wrong before (and the input arguments weren't being used?), since the input argument `Ref` wasn't being substituted with the actual input.

danking replied:
Is `cxx.Compile` actually called anywhere?

catoverdrive replied:
yep--in SparkPipeline.execute and also in assertEvalsTo->nativeExecute.

danking commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -1,29 +1,213 @@
+      throw new cxx.CXXUnsupportedOperation("IR nodes with TableIR children must be defined explicitly")
+
+    case node if node.children.exists( _.isInstanceOf[MatrixIR] ) =>
+      throw new cxx.CXXUnsupportedOperation("MatrixIR nodes must be lowered to TableIR nodes separately.")
```
Include `node` in the message?

danking commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -1,29 +1,213 @@
+        Map((rows.name, lowered) +: globals.value.stages.toSeq: _*),
+        MakeStruct(FastIndexedSeq(
+          "rows" -> ArrayFlatMap(rows, elt, Ref(elt, lowered.body.typ)),
+          "global" -> lowered.globals.head.value.body)))
```
`lowered.globals.head` appears in a couple places. Why is it safe to ignore the rest of it? What is `lowered.globals.head`? what are the tail?

catoverdrive replied:
`globals.head` is the globals of the current table. the rest of it is basically any other global information that we might want to include (e.g. the globals of the right table in a left-joined table). I thought it was documented, but maybe not--i can leave a note.

danking commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -54,9 +238,26 @@ class LowerTableIR {
+      loweredChild.copy(body = ArrayMap(loweredChild.body, row.name, Subst(newRow, env)))
+
+    case node =>
+      throw new cxx.CXXUnsupportedOperation("undefined")
```
include `node` in the message?

catoverdrive commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -1,29 +1,213 @@
+    val globsBC = RegionValue.toBytes(gEncoder, region, rvb2.end())
+    val gDecoder = codecSpec.buildDecoder(gType.physicalType, gType.physicalType)
+
+    val env = Env[IR](("context", In(1, contextType)) +: globals.map(b => b.name -> GetField(In(0, gType), b.name)): _*)
```
yes, it should (we'd treat it the same as we'd treat 'row and 'global). can generate a uuid ion the meantime though.

catoverdrive commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -1,29 +1,213 @@
+      if (typ.isFieldDefined(region, off, 0))
+        rvb2.addRegionValue(typ.types(0), region, typ.loadField(region, off, 0))
+      else
+        rvb2.setMissing()
```
ah, nice.

danking commented on hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala:
```diff
@@ -1,29 +1,199 @@
 
 case class SparkStage(
-  globals: List[Binding],
+  globals: List[SparkBinding], // globals.head is the globals of the current table; globals.tail can be any other global information that might be needed.
```
I really dislike comments when code structure suffices...

Could we do something like:

```scala
case class SparkStage(
    theRealGlobals: SparkBinding,
    otherGlobals: List[SparkBinding],
    ...
) {
    val globals = theRealGlobals +: otherGlobals
```

for both this and the local table ir? This enforces the non-emptiness of `globals` by construction as well as giving us obviously-non-error-causing ways to get the head and tail.
